### PR TITLE
passenger: provide dynamic nginx module

### PR DESCRIPTION
**Pull request by CamJN**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

I realize it's a bit weird to have nginx as a resource and build dependency, but this way there's less maintenance work keeping the nginx and passenger formulae in sync wrt the configuration flags passed to nginx. You just have to keep the resource version in sync for the dynamic module to work. And this way people can still use nginx with passenger without any options on the formula.

### CODE REVIEW

**Comment by javian** [comment]
@camJN just a thought, how about adding the missing pieces that's in the tarball to the nginx formula instead ? That way we don't have to track the resource which raises a concern about maintainability in the long term.

**Comment by CamJN** [comment]
@javian I'm sorry, but I don't quite follow, I think you're suggesting to build the passenger module in the nginx formula and then ship it there? If so, I don't think Passenger is quite popular enough to warrant everyone getting it, since options are dead.

**Comment by CamJN** [comment]
Or do you mean I should keep the source of nginx around in the nginx formula, and then just depend on nginx being installed as I do? Because if that's it, then there was a big to-do about keeping the rust source for racer's sake. @MikeMcQuaid is that still the opinion?

**Comment by javian** [comment]
@CamJN it was the latter I was thinking about since the reason you resource the nginx source is that you need access to stuff that's not in the bottle already.

**Comment by CamJN** [comment]
Another option, though previously rejected is to use something like this:
```
resource "nginx" do
  url Formula["nginx"].stable.url
  sha256 Formula["nginx"].stable.checksum.hexdigest
end
```

Removes the need to sync them, just requires rebuilding passenger when nginx updates (I think that happens anyway when there's a mandatory dependency?)

**Comment by javian** [comment]
@camjn yes, that would work. The problem is that we want reproducible isolated builds of each formula so anything that depends on something dynamic like that won't cut it as far as I know.

**Comment by javian** [comment]
Looking at it again I'm wondering if the generation of this extension isn't better suited inside the nginx formula using passenger as a `:build` dependency. Hmmm.

**Comment by CamJN** [comment]
Well, in that case I'd need someone to say it's ok to ship the Passenger module with Nginx. It's 2.5 MB and requires Passenger to be installed too, in order to actually work. But being a dynamic module, it can be ignored unless you want it.

**Comment by javian** [comment]
That's similar to what we did the in the PHP Formula https://github.com/Homebrew/homebrew-core/blob/master/Formula/php.rb#L13 where the apache module is built but apache httpd is not installed by default so that the people who want to use it can. Unless it has an impact on the working install of nginx then I don't see why this wouldn't work for passenger as well.

**Comment by MikeMcQuaid** [comment]
> Nginx's source is 1.5 MB total.

This seems fine to just ship it in the formula, then.

**Comment by CamJN** [comment]
So first, apparently when I extracted the nginx source last time, something got messed up, and I only got part of the tarball, so I was wrong about the size. The total size of the source is 6.8MB, I'm sorry about that, I really don't know what happened there, I checked my shell history and I used the same `gunzip` followed by `tar -xf` and then measure the size with `du -hc` routine both times, and I have plenty of disk space. That said, the gzipped tarball fetched by brew is 1.0MB, and recreating a tarball of the extracted source and gzipping gives me 991KB, so it might be ok to store it that way if 6.8MB is too much.

If nginx's source is going to be kept in order to facilitate compiling dynamic modules, would it be ok to create a pair of directories to include into the nginx conf, one at the top level to include load_module statements, and one in the http scope to set global conf for modules?

**Comment by MikeMcQuaid** [comment]
> That said, the gzipped tarball fetched by brew is 1.0MB, and recreating a tarball of the extracted source and gzipping gives me 991KB, so it might be ok to store it that way if 6.8MB is too much.

gzipping (or even xzipping) the tarball seems like a good call 👍

**Comment by javian** [comment]
@camjn tar also supports unzipping with the `-z` option so you don't have to use gunzip in case you want to save a step next time

**Comment by CamJN** [comment]
I've added a commit that uses the source from nginx (i didn't recreate the tarball as it's infinitely faster to just copy it into place, and it's already gzipped and only 12KB different from a re-tarred version.

File: Formula/passenger.rb
Comment (MikeMcQuaid) [inline Formula/passenger.rb]: This seems pretty hacky/brittle to me. It's looking a lot like it's not intended to be able to build a nginx passenger module like this after the fact as this doesn't guarantee the same args/flags/compiler are used.

CC @FooBarWidget for thoughts on the best approach to take here.

Comment (CamJN) [reply Formula/passenger.rb]: We ship the passenger nginx module for recent ubuntu distros in a similar way. The Ubuntu repo provides the nginx and we build a compatible dynamic module and ship that. As long as configure is called the same way, the make step should use the same flags.

Comment (MikeMcQuaid) [reply Formula/passenger.rb]: The "should" worries me a bit here. I don't think this would be widely used enough to warrant being part of nginx itself. I'll make some more inline suggestions that may make this more robust.

> So what happends when nginx would be updated ? Would the module stop working?

Yes. A `test do` will need to be added to verify this and ensure it is rebuild.

Comment (javian) [reply Formula/passenger.rb]: I haven't checked the size of the module but I doubt it can be massive but might make a large percentage of nginxs wooping 1.4MB . It seems like its practically a two line addition to the nginx formula (except the test) which would be less than adding the source files (which also consumes space possibly beyond the since of the dynamic module).

Comment (MikeMcQuaid) [reply Formula/passenger.rb]: @javian I'm more concerned about the built-time dependencies and that `nginx` will have to depend on `passenger` which feels a bit odd.

Comment (MikeMcQuaid) [reply Formula/passenger.rb]: But I'm perhaps being irrational here and would welcome thoughts from other @Homebrew/maintainers

File: Formula/passenger.rb
Comment (CamJN) [reply Formula/passenger.rb]: I agree it should be but it's one of the most common mistakes people make when installing passenger on other systems, they install the module but not the webserver itself. I can take it out but I put it there based on experience.

File: Formula/passenger.rb
Comment (CamJN) [reply Formula/passenger.rb]: Because I install the man pages into the proper directory now, at least on my system the old way wasn't working when you actually tried to run `man passenger-config` etc.

File: Formula/passenger.rb
Comment (CamJN) [reply Formula/passenger.rb]: Just out of curiosity can there be multiple `test do` blocks in a formula or should I just extend the existing one?

Comment (MikeMcQuaid) [reply Formula/passenger.rb]: Extend the existing one, thanks @CamJN.

I think the approach I've described in the PR comments of making it as easy as possible to build an nginx module by providing the source and e.g. configure args so that the passenger formula doesn't need to do much would be acceptable 👍

File: Formula/passenger.rb
Comment (MikeMcQuaid) [reply Formula/passenger.rb]: Actually, alternatively: `--strip-components=` for `tar` will avoid needing to do this.

File: Formula/passenger.rb
Comment (sjackman) [reply Formula/passenger.rb]: `passenger` could depend on `nginx`, not just `:build` but at run time too. My apologies if someone has already suggested this option. I haven't read back through the history.

Comment (CamJN) [reply Formula/passenger.rb]: @sjackman Passenger doesn't (always) need nginx, you can use it with apache instead (modulo apple currently doing stupid things with SIP in mojave), those people shouldn't need to install nginx if they don't want to use it.

**Comment by CamJN** [comment]
Ok, I pushed a new set of changes that take the feedback into account.

File: Formula/nginx.rb
Comment (MikeMcQuaid) [inline Formula/nginx.rb]: How much of a difference in size do the `excludes` make? I think it's probably worth only excluding stuff that's making a meaningful difference in size e.g. images just to avoid pain in future if stuff in the tarball moves around.

Also, at this point the `buildpath` already contains a complete copy of the source so it's probably just worth compressing that and passing the `--exclude` there instead rather than uncompressing the same data twice.

File: Formula/nginx.rb
Comment (MikeMcQuaid) [inline Formula/nginx.rb]: Specify the full output path instead of doing the `.install` below. You can also avoid the `--exclude` that way. Let's also go for `.tar.xz` as it seems more widely used.

File: Formula/nginx.rb
Comment (MikeMcQuaid) [inline Formula/nginx.rb]: `(share/"src/configure_args.txt").write args.join("\n")`

File: Formula/passenger.rb
Comment (MikeMcQuaid) [inline Formula/passenger.rb]: `args = (Formula["nginx"].opt_share/"src/args.txt").read.split("\n")`

File: Formula/passenger.rb
Comment (MikeMcQuaid) [inline Formula/passenger.rb]: Is it intentional that this variable is used twice with difference values?

File: Formula/passenger.rb
Comment (MikeMcQuaid) [reply Formula/passenger.rb]: Ok, can be left, thanks.

File: Formula/passenger.rb
Comment (MikeMcQuaid) [inline Formula/passenger.rb]:
```ruby
# to build nginx module
depends_on "nginx" => [:build, :test]
```

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

File: Formula/passenger.rb
Comment (CamJN) [reply Formula/passenger.rb]: Yes, one is the build time value, the other is the runtime value.

File: Formula/nginx.rb
Comment (CamJN) [reply Formula/nginx.rb]: The excludes make very little difference to the size of the resulting `tar.xz`, there are no images.

File: Formula/passenger.rb
Comment (CamJN) [reply Formula/passenger.rb]: done

File: Formula/passenger.rb
Comment (CamJN) [reply Formula/passenger.rb]: done

File: Formula/nginx.rb
Comment (CamJN) [reply Formula/nginx.rb]: done

File: Formula/nginx.rb
Comment (CamJN) [reply Formula/nginx.rb]: done

File: Formula/passenger.rb
Comment (MikeMcQuaid) [inline Formula/passenger.rb]: Are you sure `--strip-components` is still required here given how the tarball is created above? No worries f so, just wanted to check.

File: Formula/passenger.rb
Comment (MikeMcQuaid) [inline Formula/passenger.rb]: This is failing as it's trying to write to places it cannot: https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/37812/version=mojave/testReport/junit/brew-test-bot/mojave/test_passenger/

Comment (CamJN) [reply Formula/passenger.rb]: That's weird, as the `proxy_temp_path` and `error_log` paths are both set to sub-directories of the `testpath` in the config passed to nginx.

It also passes on my system, which might mean that the sandbox isn't active when I run `brew test`? Is there a way to turn that on locally so I can try and figure this out?

File: Formula/passenger.rb
Comment (CamJN) [reply Formula/passenger.rb]: `tar` will try to change the last-modified time on `pwd` if I leave it out; if that's ok, then it can be left out.

Comment (MikeMcQuaid) [reply Formula/passenger.rb]: Have no preference so let's leave as-is 👍

File: Formula/passenger.rb
Comment (MikeMcQuaid) [reply Formula/passenger.rb]: Not sure of anything beyond the test failure I'm afraid.

Comment (MikeMcQuaid) [reply Formula/passenger.rb]: Oh, it's maybe because this should be "#{bin}/nginx` too.

Comment (MikeMcQuaid) [reply Formula/passenger.rb]: (or at least a fully specified path)

Comment (CamJN) [reply Formula/passenger.rb]: Actually i just remembered a [passenger issue](https://github.com/phusion/passenger/issues/2075) that may be related I'll look into that and get back to you.

Comment (CamJN) [reply Formula/passenger.rb]: Um in the test `#{bin}/nginx` gives `/usr/local/Cellar/passenger/6.0.1/bin/nginx`, which isn't right. Is there a different helper for "brew home bin"?

Comment (MikeMcQuaid) [reply Formula/passenger.rb]: @CamJN Sorry, you're right, I got confused. `#{Formula["nginx"].opt_bin}/nginx` is what you want.

Comment (CamJN) [reply Formula/passenger.rb]: Ok things are looking good. I've merged the patch upstream: https://github.com/phusion/passenger/commit/7d4251bb4aec5cef6896f6d1905a30e99d21befe and will release 6.0.2 soon with it included. I'm happy with this, hope it can move forward.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
@CamJN Great work here. Sorry for all the back and forth but I think we've reached a good resolution for everyone 👍 

CC @denji who may be interested in the fact that the `nginx` formula includes the source and configure arguments.

**Comment by denji** [comment]
`buildout/nginx_dynamic` -> `modules` (dynamic modules) this path should be much better for compatibility with official nginx documentation.

**Comment by MikeMcQuaid** [comment]
> `buildout/nginx_dynamic` -> `modules` (dynamic modules) this path should be much better for compatibility with official nginx documentation.

@denji good idea, thanks. Could you submit a PR?

**Pull request closed without merging**
